src/plugins/discourse/mirror.js

```diff
@@ -0,0 +1,242 @@
+// @flow
+
+import type Database from "better-sqlite3";
+import stringify from "json-stable-stringify";
+import dedent from "../../util/dedent";
+import {
+  type DiscourseInterface,
+  type TopicId,
+  type PostId,
+  type Topic,
+  type Post,
+} from "./fetch";
+
+// The version should be bumped any time the database schema is changed,
+// so that the cache will be properly invalidated.
+const VERSION = "discourse_mirror_v1";
+
+/**
+ * Mirrors data from the Discourse API into a local sqlite db.
+ *
+ * This class allows us to persist a local copy of data from a Discourse
+ * instance. We have it for reasons similar to why we have a GraphQL mirror for
+ * GitHub; it allows us to avoid re-doing expensive IO every time we re-load
+ * SourceCred. It also gives us robustness in the face of network failures (we
+ * can keep however much we downloaded until the fault).
+ *
+ * As implemented, the Mirror will never update already-downloaded content,
+ * meaning it will not catch edits or deletions. As such, it's advisable to
+ * replace the cache periodically (perhaps once a week or month). We may
+ * implement automatic cache invalidation in the future.
+ *
+ * Each Mirror instance is tied to a particular server. Trying to use a mirror
+ * for multiple Discourse servers is not permitted; use separate Mirrors.
+ */
+export class Mirror {
+  +_db: Database;
+  +_fetcher: DiscourseInterface;
+
+  /**
+   * Construct a new Mirror instance.
+   *
+   * Takes a Database, which may be a pre-existing Mirror database. The
+   * provided DiscourseInterface will be used to retrieve new data from Discourse.
+   *
+   * A serverUrl is required so that we can ensure that this Mirror is only storing
+   * data from a particular Discourse server.
+   */
+  constructor(db: Database, fetcher: DiscourseInterface, serverUrl: string) {
+    if (db == null) throw new Error("db: " + String(db));
+    this._db = db;
+    this._fetcher = fetcher;
+    if (db.inTransaction) {
```

Comment:
paging @wchargin

Reply:
The choice by `better-sqlite3` to provide a synchronous API makes a lot
of sense; they describe the reasoning in their README and in some issues
on their repository. This was definitely a factor in choosing this
library in the first place.

The initialization should really be quite cheap—it’s doing a constant
number of operations on constant-sized data. I don’t see any reason to
make the whole API async at this point.

> having multiple instances

Not sure what you mean by this? You can of course create multiple
`Mirror` objects against the same backing database (as long as only one
mutates it), though not against the same DB _connection_, as specified
in the docs.

> skipping the init when it's in a predictable state

True, but not sure why we would do this. As mentioned above, this should
be a cheap operation. If you’re reinitializing databases in a hot loop…
don’t do that. The benefits that we get from error checking are real.

For context, we’ve seen from the GitHub fetching pipeline that the
SQLite operations are nowhere _near_ a bottleneck. They complete roughly
immediately compared to the GitHub operations, which take multiple
seconds each.

> deferring I/O during a DI stage

Not sure what you mean by this? You could dependency-inject either an
alternate implementation of the `Database` connection or an alternate
implementation of the mirror itself with no problems.

Reply:
> The initialization should really be quite cheap—it’s doing a constant
> number of operations on constant-sized data. I don’t see any reason to
> make the whole API async at this point.

Yes at this point it's a bit too early to worry about the performance of this. Having 20 plugins do the same thing in the future though there may be some value in having a separate function that can be made async if necessary. While the constructor doesn't have this option.

> > having multiple instances
> 
> Not sure what you mean by this? You can of course create multiple
> `Mirror` objects against the same backing database (as long as only one
> mutates it), though not against the same DB _connection_, as specified
> in the docs.

> If you’re reinitializing databases in a hot loop… don’t do that.

Is what I meant by the multiple instance point. It's not impossible, but undesirable because of this choice. The performance impact of the init would mean you avoid constructing a lot of instances. Which in theory shouldn't be an issue as the `new` operation is pretty fast in comparison.

> > deferring I/O during a DI stage
> 
> Not sure what you mean by this? You could dependency-inject either an
> alternate implementation of the `Database` connection or an alternate
> implementation of the mirror itself with no problems.

What I mean is, a lot of dependency injection patterns happen on application startup and don't assume I/O heavy constructors. When you think of a Mirror instance as simply a repository instance with the correct database handles and config values passed in that should be cheap enough to repeat for 100 repositories without impacting startup time too much.

Imagine a large project which has 20 `sourcecred scores <repo>` calls in a shell script loop and 5 plugins which have initialization SQL transactions. To avoid the unnecessary I/O here you'll either want to have lazy-loading DI or just very lightweight classes during the DI step is it doesn't matter we're not using them.

Reply:
> The benefits that we get from error checking are real.

In favour of doing things this way, if you were to view this as a runtime check for bad arguments being passed to the constructor, the constructor would make sense.

However I think a SQL transaction with side-effects is not quite comparable to something like a runtime check for required arguments or pattern matching some string looks like an email address.

Reply:
If you want to have the benefits of both. Why not use a factory pattern? IPFS [recently implemented](https://github.com/ipfs/js-ipfs/issues/2192) something similar:

```javascript
// From:
const IPFS = require('ipfs')
const node = new IPFS()
await node.ready // A separate init function

// To:
const IPFS = require('ipfs')
const node = await IPFS.create() // Init as part of the factory function
```

This way you could have a lightweight constructor without side-effects for "risky" usage without an init call, but use the async function as a convention to include the init step as well.